src/electionguard/precompute_buffers.cpp

```diff
@@ -0,0 +1,243 @@
+#include <electionguard/precompute_buffers.hpp>
+
+#include <array>
+#include <cstdint>
+#include <electionguard/constants.h>
+#include <electionguard/export.h>
+#include <iomanip>
+#include <iostream>
+#include <memory>
+
+using std::begin;
+using std::copy;
+using std::end;
+using std::unique_ptr;
+using std::make_unique;
+using std::begin;
+using std::copy;
+using std::end;
+
+namespace electionguard
+{
+    void Triple::generateTriple(const ElementModP &publicKey)
+    {
+        // generate a random rho
+        rho = rand_q();
+        g_to_rho = g_pow_p(*rho);
+        pubkey_to_rho = pow_mod_p(publicKey, *rho);
+    }
+
+    void Triple::set(unique_ptr<ElementModQ> in_rho, unique_ptr<ElementModP> in_g_to_rho,
+                     unique_ptr<ElementModP> in_pubkey_to_rho)
+    {
+        rho = move(in_rho);
+        g_to_rho = move(in_g_to_rho);
+        pubkey_to_rho = move(in_pubkey_to_rho);
+    }
+
+    void Quadruple::generateQuadruple(const ElementModP &publicKey)
+    {
+        // generate a random sigma and rho
+        sigma = rand_q();
+        rho = rand_q();
+        g_to_rho = g_pow_p(*rho);
+        g_to_sigma_mult_by_pubkey_to_rho = mul_mod_p(*g_pow_p(*sigma), *pow_mod_p(publicKey, *rho));
+    }
+
+    void Quadruple::set(unique_ptr<ElementModQ> in_rho, unique_ptr<ElementModQ> in_sigma,
+                        unique_ptr<ElementModP> in_g_to_rho,
+                        unique_ptr<ElementModP> in_g_to_sigma_mult_by_pubkey_to_rho)
+    {
+        rho = move(in_rho);
+        sigma = move(in_sigma);
+        g_to_rho = move(in_g_to_rho);
+        g_to_sigma_mult_by_pubkey_to_rho = move(in_g_to_sigma_mult_by_pubkey_to_rho);
+    }
+
+    void TripleEntry::set(unique_ptr<Triple> triple)
+    {
+        unique_ptr<ElementModQ> p_rho = triple->get_rho();
+        unique_ptr<ElementModP> p_g_to_rho = triple->get_g_to_rho();
+        unique_ptr<ElementModP> p_pubkey_to_rho = triple->get_pubkey_to_rho();
+        uint64_t *p_rho_u64 = p_rho->get();
+        uint64_t *p_g_to_rho_u64 = p_g_to_rho->get();
+        uint64_t *p_pubkey_to_rho_u64 = p_pubkey_to_rho->get();
+        copy(p_rho_u64, p_rho_u64 + MAX_Q_LEN, rho);
+        copy(p_g_to_rho_u64, p_g_to_rho_u64 + MAX_P_LEN, g_to_rho);
+        copy(p_pubkey_to_rho_u64, p_pubkey_to_rho_u64 + MAX_P_LEN, pubkey_to_rho);
+    }
+
+    unique_ptr<Triple> TripleEntry::get_triple()
+    {
+        unique_ptr<Triple> result = make_unique<Triple>();
+        
+        unique_ptr<ElementModQ> temp_rho = make_unique<ElementModQ>(rho);
+        unique_ptr<ElementModP> temp_g_to_rho = make_unique<ElementModP>(g_to_rho);
+        unique_ptr<ElementModP> temp_pubkey_to_rho = make_unique<ElementModP>(pubkey_to_rho);
+
+        result->set(move(temp_rho), move(temp_g_to_rho), move(temp_pubkey_to_rho));
+        
+        return result;
+    }
+    
+    void QuadrupleEntry::set(unique_ptr<Quadruple> quadruple)
+    { 
+        unique_ptr<ElementModQ> p_sigma = quadruple->get_sigma();
```

Comment:
im a bit confused by what's going on in here, it looks like calls like `quadruple->get_sigma()` create a unique pointer with a move operation, and then line 93 below (`p_sigma->get()` unwraps the pointer, and then there's an explicit copy operation happening. is this overwriting memory that was already populated? is ther something im missing that necessitates using the explicit copy operation?

did you happen to notice the [clone()](https://github.com/microsoft/electionguard-cpp/blob/main/src/electionguard/group.cpp#L241) methods on most objects? that might be a bit easier if you can get away with it.

Reply:
Thanks for the comment! I've implemented clone and done some cleanup.